### PR TITLE
[SPARK-27259][CORE] Allow setting -1 as length for FileBlock

### PR DESCRIPTION
**Pull request by praneetsharma**
### What changes were proposed in this pull request?

This PR aims to update the validation check on `length` from `length >= 0` to `length >= -1` in order to allow set `-1` to keep the default value.

### Why are the changes needed?

At Apache Spark 2.2.0, [SPARK-18702](https://github.com/apache/spark/pull/16133/files#diff-2c5519b1cf4308d77d6f12212971544fR27-R38) adds `class FileBlock` with the default `length` value, `-1`, initially.

There is no way to set `filePath` only while keeping `length` is `-1`.
```scala
  def set(filePath: String, startOffset: Long, length: Long): Unit = {
     require(filePath != null, "filePath cannot be null")
     require(startOffset >= 0, s"startOffset ($startOffset) cannot be negative")
     require(length >= 0, s"length ($length) cannot be negative")
     inputBlock.set(new FileBlock(UTF8String.fromString(filePath), startOffset, length))
   }
```

For compressed files (like GZ), the size of split can be set to -1. This was allowed till Spark 2.1 but regressed starting with spark 2.2.x. Please note that split length of -1 also means the length was unknown - a valid scenario. Thus, split length of -1 should be acceptable like pre Spark 2.2.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

This is updating the corner case on the requirement check. Manually check the code.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @praneetsharma .
Could you give us the valid reference for the following claim?
> Please note that split length of -1 also means the length was unknown - a valid scenario.

File: core/src/main/scala/org/apache/spark/rdd/InputFileBlockHolder.scala [hunk 1]
```diff
@@ -76,7 +76,6 @@ private[spark] object InputFileBlockHolder {
   def set(filePath: String, startOffset: Long, length: Long): Unit = {
     require(filePath != null, "filePath cannot be null")
     require(startOffset >= 0, s"startOffset ($startOffset) cannot be negative")
-    require(length >= 0, s"length ($length) cannot be negative")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/rdd/InputFileBlockHolder.scala]: If `-1` is valid, why not `length >= -1` instead of `length >= 0`?

Comment (praneetsharma) [reply core/src/main/scala/org/apache/spark/rdd/InputFileBlockHolder.scala]: Hi @dongjoon-hyun 

The file InputFileBlockHolder contains the following statement stating that -1 can be returned as length if the length is unknown:
```
  /**
   * Returns the length of the block being read, or -1 if it is unknown.
   */
  def getLength: Long = inputBlock.get().get().length
```

Yes, I can modify the validation statement to be `length >= -1`. In that case, what can be a good message? Maybe - "length ($length) cannot be smaller than -1"?

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/rdd/InputFileBlockHolder.scala]: First of all, it's `private` to Spark. You should not depend on the private implementation of Apache Spark. There is no guarantee for that and this is not a regression of Apache Spark.

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/rdd/InputFileBlockHolder.scala]: Yes. Please update the PR with `length >= -1`. For the message, that message looks okay. Then, probably, it's okay.
> Yes, I can modify the validation statement to be length >= -1. In that case, what can be a good message? Maybe - "length ($length) cannot be smaller than -1"?

**Comment by dongjoon-hyun** [comment]
ok to test

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @praneetsharma .
Merged to `master/3.0-preview/2.4`.

**Pull request closed without merging**